libs/domain/cart/coupon/coupon.component.ts

```
@@ -0,0 +1,128 @@
+            <oryx-icon
+              type=${IconTypes.Close}
+              class="clear-icon"
+              appearance=${ClearIconAppearance.Hover}
+              @click=${(): void => this.clear()}
+            ></oryx-icon>
+          </oryx-input>
+          <oryx-button
+            .type="${ButtonType.Outline}"
+            .color="${ColorType.Neutral}"
+            .size=${ButtonSize.Md}
+            @click=${this.onSubmit}
+          >
+            ${this.i18n('coupon.apply')}
+          </oryx-button>
+        </section>
+      </oryx-layout>
+
+      <ul>
+        ${repeat(
+          this.$coupons(),
+          (coupon) => coupon.code,
+          (coupon) => {
+            let expirationDateTime = '';
+            this.localeService
+              .formatDate(coupon.expirationDateTime)
+              .pipe(
+                tap((date) => {
+                  expirationDateTime = date ?? '';
+                })
+              )
+              .subscribe();
+            return html`
+              <li>
+                <oryx-icon type="done"></oryx-icon>
+                <span class="code">${coupon.code} </span>
+                <span class="name"
+                  >${coupon.displayName}
+                  <span>(valid till ${expirationDateTime})</span>
+                </span>
+              </li>
+            `;
+          }
+        )}
+      </ul>
+    `;
+  }
+
+  protected clear(): void {
+    this.coupon!.value = '';
+  }
+
+  protected onSubmit(): void {
+    this.hasError = false;
+
+    if (this.coupon?.value === '') {
+      this.hasError = true;
+
+      return;
+    }
+
+    this.cartService.addCoupon({ code: this.coupon?.value ?? '' }).subscribe({
+      next: () => {
+        this.notificationService.push({
+          type: AlertType.Success,
+          content: `'${this.coupon?.value}' successfully applied`,
+        });
+
+        this.coupon!.value = '';
+      },
+      error: () => {
+        this.hasError = true;
```

I believe error should be handle in query, and we should observe it, instead of keeping local state

---

In this case. we need it as local state to control updates to the element, due to changes when submitting and to show/hide error messages more dynamically.